lib/pos_mempool.go

```diff
@@ -0,0 +1,446 @@
+	for it.Next() {
+		if txn, ok := it.Value(); ok {
+			txns = append(txns, txn)
+		}
+	}
+	return txns
+}
+
+// FeeTimeIterator is an iterator over the transactions in a FeeBucket. The iterator goes through all transactions
+// as ordered by Fee-Time.
+type FeeTimeIterator struct {
+	// feeBucketIterator is an iterator over the TimeBuckets in the FeeBucket.
+	feeBucketIterator treeset.Iterator
+	// timeBucketIterator is an iterator over the transactions in the current TimeBucket.
+	timeBucketIterator treeset.Iterator
+	// initialized is set to true when the iterator is pointing at a transactions.
+	initialized bool
+}
+
+// Next moves the FeeTimeIterator to the next transaction. It returns true if the iterator is pointing at a transaction
+// after the move, and false otherwise.
+func (fti *FeeTimeIterator) Next() bool {
+	timeBucketNext := false
+
+	// If the iterator is not initialized, we will make it advance the feeBucketIterator and timeBucketIterator.
+	if !fti.initialized {
+		timeBucketNext = false
+		fti.initialized = true
+	} else {
+		// If the iterator is initialized, we will first see if there are more transactions in the current TimeBucket.
+		// We do this by advancing the timeBucketIterator and checking if it is pointing at a transaction.
+		timeBucketNext = fti.timeBucketIterator.Next()
+	}
+
+	if !timeBucketNext {
+		// If there are no more transactions in the current TimeBucket, we will advance the feeBucketIterator.
+		if !fti.feeBucketIterator.Next() {
+			// If there are no more TimeBuckets in the FeeBucket, we are done.
+			return false
+		}
+
+		// If there are more TimeBuckets in the FeeBucket, we will advance the timeBucketIterator to the first transaction
+		// in the new TimeBucket.
+		if timeBucketVal, ok := fti.feeBucketIterator.Value().(*TimeBucket); ok {
+			fti.timeBucketIterator = timeBucketVal.txnsSet.Iterator()
+			// We will advance the timeBucketIterator to the first transaction in the new TimeBucket.
+			if !fti.timeBucketIterator.Next() {
+				// This shouldn't happen, but if it does, we've reached the end of the FeeBucket.
+				return false
+			}
+		} else {
+			return false
+		}
+	}
+	return true
+}
+
+// Value returns the transaction that the iterator is pointing at.
+func (fti *FeeTimeIterator) Value() (*MempoolTx, bool) {
+	if !fti.initialized {
+		return nil, false
+	}
+
+	if txn, ok := fti.timeBucketIterator.Value().(*MempoolTx); ok {
+		return txn, true
+	} else {
+		return nil, false
+	}
+}
+
+// ========================
+//	TimeBucket
+// ========================
+
+// TimeBucket is a data structure storing MempoolTx with similar fee rates. It has an index, which determines the
+// range of fee rates that are accepted by the TimeBucket.
+type TimeBucket struct {
+	// feeMin is the minimum fee rate (inclusive) accepted by the TimeBucket, in nanos per KB.
+	feeMin uint64
+	// feeMax is the maximum fee rate (inclusive) accepted by the TimeBucket, in nanos per KB.
+	feeMax uint64
```

Comment:
Nit, update these fields so their units are contextualized in the name
```suggestion
	minFeeDESONanosPerKB uint64
	// feeMax is the maximum fee rate (inclusive) accepted by the TimeBucket, in nanos per KB.
	maxFeeDESONanosPerKB uint64
```